packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```diff
@@ -0,0 +1,633 @@
+        _;
+    }
+    ///@dev prevent race conditions when rebalancing
+    bool private isRebalancing;
+
+    modifier nonReentrantRebalance() {
+        require(!isRebalancing, "Rebalance in progress");
+        isRebalancing = true;
+        _;
+        isRebalancing = false;
+    }
+
+    /***************************************************************************
+                                  Router Functions
+    ***************************************************************************/
+    function addLiquidityProportional(
+        address pool,
+        uint256[] memory maxAmountsIn,
+        uint256 exactBptAmountOut,
+        bool wethIsEth,
+        bytes memory userData
+    ) external payable saveSender returns (uint256[] memory amountsIn) {
+        // Do addLiquidity operation - BPT is minted to this contract.
+        amountsIn = _addLiquidityProportional(
+            pool,
+            msg.sender,
+            address(this),
+            maxAmountsIn,
+            exactBptAmountOut,
+            wethIsEth,
+            userData
+        );
+        address[] storage lps = liquidityProviders[pool];
+        lps.push(msg.sender);
+        uint256[] storage amounts = amountTokens[pool][msg.sender];
+        for (uint256 i = 0; i < amounts.length; i++) {
+            amounts[i] += amountsIn[i];
+        }
+        isStillLp[pool][msg.sender] = true;
+        emit LiquidtyProviderAdded(pool, msg.sender);
+    }
+
+    function removeLiquidityProportional(
+        uint256 tokenId,
+        address pool,
+        uint256 exactBptAmountIn,
+        uint256[] memory minAmountsOut,
+        bool wethIsEth
+    ) external payable saveSender returns (uint256[] memory amountsOut) {
+        // Do removeLiquidity operation - tokens sent to msg.sender.
+        amountsOut = _removeLiquidityProportional(
+            pool,
+            msg.sender,
+            msg.sender,
+            exactBptAmountIn,
+            minAmountsOut,
+            wethIsEth,
+            ""
+        );
+        uint256[] memory lpBptAmount = amountTokens[pool][msg.sender];
+
+        uint256 length = amountsOut.length;
+        for (uint256 i = 0; i < length; i++) {
+            lpBptAmount[i] -= amountsOut[i];
+        }
+
+        bool hasRemovedAll = true;
+
+        for (uint256 i = 0; i < length; i++) {
+            if (lpBptAmount[i] != 0) {
+                hasRemovedAll = false;
+            }
+        }
+
+        isStillLp[pool][msg.sender] = hasRemovedAll;
+    }
+
+    /***************************************************************************
+                                  Hook Functions
+    ***************************************************************************/
+    ///@inheritdoc IHooks
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public override onlyVault returns (bool) {
+        ///@dev removing for testing
```

Comment:
The check would be that the factory is the WeightedPoolFactory - very important, or else it wouldn't work at all - and then that the pool is from that factory. Should be able to test with a real weighted pool

Reply:
Yup so for the longest time I couldn't find a example for weighted pool test as for all the example it was done otherwise. So I don't know what I thought of when I tried to test for a normal pool (a little sleep deprived I guess) and commented this line of code out.

I finally looked at Weighted pool test itself and got how to set up weighted pool itself. So removing this line would work and I could also put in a test for wrong factory check